modules/Movecraft/src/main/java/net/countercraft/movecraft/towny/TownyCompatManager.java

```diff
@@ -0,0 +1,146 @@
+        final Set<TownBlock> townBlocks = new HashSet<>();
+        final Craft craft = event.getCraft();
+        final TownyWorld townyWorld = TownyUtils.getTownyWorld(craft.getW());
+        if (!Settings.TownyBlockMoveOnSwitchPerm)
+            return;
+        for (MovecraftLocation ml : event.getNewHitBox()) {
+
+            TownBlock townBlock = TownyUtils.getTownBlock(ml.toBukkit(event.getCraft().getW()));
+            if (townBlock == null || townBlocks.contains(townBlock)) {
+                continue;
+            }
+            if (TownyUtils.validateCraftMoveEvent(craft.getNotificationPlayer(), ml.toBukkit(craft.getW()), townyWorld)) {
+                townBlocks.add(townBlock);
+                continue;
+            }
+            Town town = TownyUtils.getTown(townBlock);
+            if (town == null)
+                continue;
+            final TownyWorldHeightLimits whLim = TownyUtils.getWorldLimits(event.getCraft().getW());
+            final Location spawnLoc = TownyUtils.getTownSpawn(townBlock);
+            if (whLim.validate(ml.getY(), spawnLoc.getBlockY())) {
+                continue;
+            }
+            event.setFailMessage(String.format(I18nSupport.getInternationalisedString("Towny - Translation Failed") + " %s @ %d,%d,%d", town.getName(), ml.getX(), ml.getY(), ml.getZ()));
+            event.setCancelled(true);
+            break;
+        }
+
+    }
+
+    @EventHandler
+    public void onCraftRotate(CraftRotateEvent event) {
+        final Set<TownBlock> townBlocks = new HashSet<>();
+        final Craft craft = event.getCraft();
+        final TownyWorld townyWorld = TownyUtils.getTownyWorld(craft.getW());
+        if (!Settings.TownyBlockMoveOnSwitchPerm)
+            return;
+        for (MovecraftLocation ml : event.getNewHitBox()) {
+
+            TownBlock townBlock = TownyUtils.getTownBlock(ml.toBukkit(event.getCraft().getW()));
+            if (townBlock == null || townBlocks.contains(townBlock)) {
+                continue;
+            }
+            if (TownyUtils.validateCraftMoveEvent(craft.getNotificationPlayer(), ml.toBukkit(craft.getW()), townyWorld)) {
+                townBlocks.add(townBlock);
+                continue;
+            }
+            Town town = TownyUtils.getTown(townBlock);
+            if (town == null)
+                continue;
+            final TownyWorldHeightLimits whLim = TownyUtils.getWorldLimits(event.getCraft().getW());
+            final Location spawnLoc = TownyUtils.getTownSpawn(townBlock);
+            if (whLim.validate(ml.getY(), spawnLoc.getBlockY())) {
+                continue;
+            }
+            event.setFailMessage(String.format(I18nSupport.getInternationalisedString("Towny - Rotation Failed") + " %s @ %d,%d,%d", town.getName(), ml.getX(), ml.getY(), ml.getZ()));
+            event.setCancelled(true);
+            break;
+        }
+    }
+
+    @EventHandler
+    public void onHitboxDetect(HitBoxDetectEvent event) {
```

Comment:
Why is this event necessary? `CraftDetectEvent` should already have the hitbox

Reply:
CraftDetectEvent cannot be cancelled